module_3/gifts/repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java

```diff
@@ -0,0 +1,323 @@
+
+	public CertificateRepositoryImpl() {
+
+	}
+
+	/**
+	 * Saves the passed certificate.
+	 * 
+	 * @param certificateModel the certificate to be saved
+	 * @return saved certificate
+	 */
+	@Override
+	@Transactional
+	public CertificateModel save(CertificateModel certificateModel) {
+		entityManager.persist(certificateModel);
+		return certificateModel;
+	}
+
+	/**
+	 * Reads certificate with passed id.
+	 * 
+	 * @param certificateId the id of certificate to be read
+	 * @return certificate with passed id
+	 */
+	@Override
+	public Optional<CertificateModel> findById(long certificateId) {
+		try {
+			return Optional.of(obtainReadByIdQuery(certificateId).getSingleResult());
+		} catch (NoResultException e) {
+			return Optional.empty();
+		}
+	}
+
+	/**
+	 * Checks whether certificate with passed id exists.
+	 * 
+	 * @param certificateId the id of tag to be checked
+	 * @return {@code true} if the the certificate with passed id already exists and
+	 *         {@code false} otherwise
+	 */
+	@Override
+	public boolean certificateExistsById(long certificateId) {
+		try {
+			obtainReadByIdQuery(certificateId).getSingleResult();
+		} catch (NoResultException e) {
+			return false;
+		}
+		return true;
+
+	}
+
+	private TypedQuery<CertificateModel> obtainReadByIdQuery(long tagId) {
+		CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+		CriteriaQuery<CertificateModel> certificateCriteria = criteriaBuilder.createQuery(CertificateModel.class);
+		Root<CertificateModel> certificateRoot = certificateCriteria.from(CertificateModel.class);
+		certificateCriteria.select(certificateRoot);
+
+		Predicate isDeletedPredicate = criteriaBuilder.equal(certificateRoot.get(CertificateModel_.isDeleted), false);
+		Predicate idPredicate = criteriaBuilder.equal(certificateRoot.get(CertificateModel_.id), tagId);
+		certificateCriteria.where(isDeletedPredicate, idPredicate);
+
+		return entityManager.createQuery(certificateCriteria);
+	}
+
+	/**
+	 * Reads certificate with passed name.
+	 * 
+	 * @param certificateName the name of certificate to be read
+	 * @return certificate with passed name
+	 */
+	@Override
+	public Optional<CertificateModel> findByName(String certificateName) {
+		try {
+			return Optional.of(obtainReadByNameQuery(certificateName).getSingleResult());
+		} catch (NoResultException e) {
+			return Optional.empty();
+		}
+	}
+
+	/**
+	 * Checks whether the certificate with passed name already exists.
+	 * 
+	 * @param certificateName the name of certificate to check
+	 * @return {@code true} if the the certificate with passed name already exists
+	 *         and {@code false} otherwise
+	 */
+	@Override
+	public boolean certificateExistsByName(String certificateName) {
+		try {
+			obtainReadByNameQuery(certificateName).getSingleResult();
+		} catch (NoResultException e) {
+			return false;
+		}
+		return true;
+	}
+
+	private TypedQuery<CertificateModel> obtainReadByNameQuery(String certificateName) {
```

Comment:
@LizavetaSinitsyna 
Could you move all private assembly requests to a separate class? This makes your code easier to understand.